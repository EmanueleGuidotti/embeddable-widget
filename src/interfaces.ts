export interface VendorInterface {
  name: string;
  departureAirport: string;
  map: string;
  features: string[];
}
export interface ParkingInterface {
  id: number;
  indoor: boolean;
  insurance: boolean;
  price: string;
}
export interface DictionaryInterface {
  bookYourParking: string;
  featuresTitle: string;
  indoorSpace: string;
  outdoorSpace: string;
  insuranceIncluded: string;
  insuranceExcluded: string;
}
export interface ModelInterface {
  vendor: VendorInterface;
  parkings: ParkingInterface[];
  dictionary: DictionaryInterface;
}
