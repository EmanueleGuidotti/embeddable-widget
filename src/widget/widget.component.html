<section class="vendor">
  <div class="vendor__toggleable">
    <p class="vendor__name"></p>
    <p class="vendor__captureTxt"></p>
    <button class="vendor__btn">open</button>
  </div>
  <div class="vendor__features">
    <div class="features__detail">
      <div class="vendor__map-container">
        <img class="vendor__map" alt="parking-map" />
      </div>
      <div class="features__list-container">
        <p class="features__title"></p>
        <ul class="features__list"></ul>
      </div>
    </div>
  </div>
</section>
<section class="parking__slots"></section>
