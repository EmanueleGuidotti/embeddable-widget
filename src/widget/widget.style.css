@keyframes open {
  0% {
    opacity: 0;
    height: 0;
  }
  100% {
    opacity: 1;
    height: 100%;
  }
}

.vendor {
  display: flex;
  flex-direction: column;
  font-size: 12px;
  background: #efefef;
  border-radius: 10px;
  padding: 0;
  box-sizing: border-box;
  font-family: Arial;
}

.vendor__name {
  font-size: 26px;
  font-weight: 600;
}

.vendor__toggleable {
  display: flex;
  flex-direction: column;
  background: #efefef;
  border-radius: 10px;
  padding: 10px 30px;
  box-sizing: border-box;
}

.vendor__btn {
  float: right;
}

.vendor__features {
  height: 0;
  overflow: hidden;
}

.vendor__features.opened {
  animation: open 2s forwards ease-in-out;
  animation-iteration-count: 1;
}

.vendor__btn {
  background: olive;
  text-transform: uppercase;
  border: none;
  border-radius: 10px;
  padding: 1em 3.3em;
}

.features__detail {
  padding: 10px 30px 25px;
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  justify-content: flex-start;
}

.vendor__map-container {
  flex: 1;
  width: 100%;
  height: auto;
  max-width: 60%;
  margin: auto;
}

.vendor__map {
  width: 100%;
  height: auto;
}

.features__list-container {
  flex: 3 0 auto;
  padding: 20px;
  padding-bottom: 0;
}

.features__list {
  padding: 0;
  list-style: none;
}

.features__list-item {
  list-style: inside;
  padding: 10px;
}

.features__title {
  font-size: 20px;
  color: olivedrab;
  margin: 0;
}

.parking__slots {
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
}

.parking__item {
  flex: 1 0 50%;
  padding: 10px;
  background: red;
  margin: 10px 0 0 0;
  border-radius: 10px;
}

.parking__item.selected {
  background: aquamarine;
  color: beige;
}

/* medium-big screen */
@media only screen and (min-width: 700px) {
  .vendor__toggleable {
    display: block;
  }
  .vendor__map-container {
    flex: 2;
  }
  .features__detail {
    flex-direction: row;
  }
  .features__list-container {
    flex: 3;
    padding: 0 20px;
  }
  .parking__slots {
    flex-direction: row;
  }
  .parking__item {
    flex: 2 0 auto;
    margin: 15px 10px;
  }
}
