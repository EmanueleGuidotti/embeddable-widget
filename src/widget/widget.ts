import html from './widget.component.html';
import './widget.style.css';
import {
  ModelInterface,
  ParkingInterface,
  VendorInterface,
  DictionaryInterface,
} from '../interfaces';

export class Widget {
  // DOM variables
  templateSlot: HTMLDivElement;
  body: HTMLBodyElement;
  parkingSlots = [] as HTMLElement[];

  // Data variables
  readonly vendor: VendorInterface;
  readonly parking: ParkingInterface[];
  readonly dictionary: DictionaryInterface;
  setParking: Function;

  // Initialization and model injection
  constructor(model: ModelInterface) {
    this.vendor = model.vendor;
    this.parking = model.parkings;
    this.dictionary = model.dictionary;
  }

  // A Callback for receiving the price info
  set parkingSelection(fn: any) {
    this.setParking = fn;
  }

  // Main initialization method, inject the widject container inside the dom
  public async init(): Promise<any> {
    // convert plain HTML string into DOM elements
    this.templateSlot = document.createElement('div');
    this.templateSlot.innerHTML = html;
    // append elements to body
    this.body = document.getElementsByTagName('body')[0];
    this.fillTheDom();
  }

  // This method will fill the widget with all the dom elements
  fillTheDom(): void {
    this.body.appendChild(this.templateSlot);
    // Create the rest
    this.createVendor();
    this.createFeatures();
    this.createParkingSlots();
  }

  // Toggle open and close of the vendor map area
  toggleOpenClose(
    button: HTMLButtonElement,
    element: HTMLDivElement,
    className: string,
  ): EventListener {
    return () => {
      element.classList.toggle(className);
      button.textContent === 'open'
        ? (button.textContent = 'close')
        : (button.textContent = 'open');
    };
  }

  // Create the parking slot and attache the click event to it
  createParkingSlot(parking: ParkingInterface): HTMLDivElement {
    let slot = document.createElement('div');
    const { id, indoor, insurance, price } = parking;
    const {
      indoorSpace,
      outdoorSpace,
      insuranceExcluded,
      insuranceIncluded,
    } = this.dictionary;
    slot.classList.add('parking__item');
    slot.setAttribute('data-id', id.toString());
    slot.setAttribute('data-indoor', indoor ? indoorSpace : outdoorSpace);
    slot.setAttribute(
      'data-insurance',
      insurance ? insuranceIncluded : insuranceExcluded,
    );
    slot.setAttribute('data-price', price);
    slot.textContent = price;
    this.attachEventToSlot(slot, parking);
    return slot;
  }

  // Click event for parking slots
  attachEventToSlot(
    parkingSlot: HTMLDivElement,
    parking: ParkingInterface,
  ): void {
    let pid: string = parkingSlot.getAttribute('data-id');
    parkingSlot.addEventListener('click', (e) => {
      parkingSlot.classList.add('selected');
      this.parkingSlots.forEach((el: HTMLDivElement, index: number) => {
        let id: string = el.getAttribute('data-id');
        if (id !== pid) el.classList.remove('selected');
      });
      this.setParking(parking);
    });
  }

  // Create the parking slot container
  createParkingSlots(): void {
    const parkingList = this.templateSlot.getElementsByClassName(
      'parking__slots',
    )[0] as HTMLElement;
    this.parking.forEach((p: ParkingInterface) => {
      let parking = this.createParkingSlot(p);
      this.parkingSlots.push(parking); // create the matrix
      parkingList.appendChild(parking);
    });
  }

  // Create the feature area
  createFeatures(): void {
    const { features } = this.vendor;
    const featuresList = this.templateSlot.getElementsByClassName(
      'features__list',
    )[0] as HTMLElement;
    for (let f of features as string[]) {
      let featureElement = document.createElement('li');
      featureElement.classList.add('features__list-item');
      featureElement.textContent = f;
      featuresList.appendChild(featureElement);
    }
    // Attach toggle to button and container
    const button = this.templateSlot.getElementsByClassName('vendor__btn')[0];
    const toggleable = this.templateSlot.getElementsByClassName(
      'vendor__features',
    )[0];
    button.addEventListener(
      'click',
      <EventListener>(
        this.toggleOpenClose(
          <HTMLButtonElement>button,
          <HTMLDivElement>toggleable,
          'opened',
        )
      ),
    );
  }

  // Create the vendor area
  createVendor(): void {
    const { name, departureAirport, map } = this.vendor;
    const { bookYourParking, featuresTitle } = this.dictionary;
    const vendorMap = this.templateSlot.getElementsByClassName(
      'vendor__map',
    )[0] as HTMLImageElement;
    vendorMap.src = map;
    this.templateSlot.getElementsByClassName(
      'vendor__name',
    )[0].textContent = name;
    this.templateSlot.getElementsByClassName(
      'vendor__captureTxt',
    )[0].textContent = bookYourParking + ' ' + departureAirport;
    this.templateSlot.getElementsByClassName(
      'features__title',
    )[0].textContent = featuresTitle;
  }
}
