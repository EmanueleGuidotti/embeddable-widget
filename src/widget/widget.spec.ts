import { ModelInterface } from '../interfaces';
const model: ModelInterface = {
  vendor: {
    name: 'ParkinGo',
    departureAirport: 'Bergamo Orio al Serio',
    map: 'http://lorempixel.com/370/390',
    features: [
      'Parcheggio ParkinGo della tipologia selezionata',
      'Navetta da/per aeroporto 24h',
      'Vigilanza e custodia del mezzo 24h',
      'Ripristino gratuito del calo di batteria della tua auto',
    ],
  },
  parkings: [
    {
      id: 123,
      indoor: false,
      insurance: false,
      price: '13.56 EUR',
    },
    {
      id: 456,
      indoor: false,
      insurance: true,
      price: '18.99 EUR',
    },
    {
      id: 243,
      indoor: true,
      insurance: true,
      price: '23.56 EUR',
    },
    {
      id: 700,
      indoor: true,
      insurance: false,
      price: '16.49 EUR',
    },
  ],
  dictionary: {
    bookYourParking: 'Prenota il tuo parcheggio a',
    featuresTitle: 'Sono inclusi i seguenti servizi:',
    indoorSpace: 'Posto auto coperto',
    outdoorSpace: 'Posto auto scoperto',
    insuranceIncluded: 'Assicurazione inclusa',
    insuranceExcluded: 'Assicurazione esclusa',
  },
};
import { Widget } from './widget';

describe('Widget class', () => {
  let widgetClass: Widget;

  beforeEach(() => {
    widgetClass = new Widget(model);
    document.body.innerHTML = '';
    widgetClass.templateSlot = document.createElement('div');
    widgetClass.templateSlot.innerHTML = `
                    <section class="vendor">
                        <div class="vendor__toggleable">
                            <p class="vendor__name"></p>
                            <p class="vendor__captureTxt"></p>
                            <button class="vendor__btn">open</button>
                        </div>
                        <div class="vendor__features">
                            <div class="features__detail">
                                <div class="vendor__map-container"><img class="vendor__map" alt="parking-map"></div>
                                <div class="features__list-container">
                                    <p class="features__title"></p>
                                    <ul class="features__list"></ul>
                                </div>
                            </div>
                        </div>
                    </section>
                    <section class="parking__slots">
                    </section>`;
    widgetClass.parkingSelection = jest.fn();
    document.body.appendChild(widgetClass.templateSlot);
  });

  it('should templateSlot created', () => {
    expect(widgetClass.templateSlot).toBeTruthy();
    expect(widgetClass.templateSlot).toBeDefined();
    // @ts-ignore
    expect(widgetClass.templateSlot).toMatchSnapshot();
  });

  describe('Events works properly', () => {
    it('should toggleOpenClose', () => {
      const button = document.getElementsByClassName(
        'vendor__btn',
      )[0] as HTMLButtonElement;
      const toggleable = document.getElementsByClassName(
        'vendor__features',
      )[0] as HTMLDivElement;
      button.addEventListener(
        'click',
        <EventListener>(
          widgetClass.toggleOpenClose(
            <HTMLButtonElement>button,
            <HTMLDivElement>toggleable,
            'opened',
          )
        ),
      );
      button.dispatchEvent(new Event('click'));
      expect(toggleable.classList.contains('opened')).toBe(true);
      expect(button.textContent).toBe('close');
      // @ts-ignore
      expect(widgetClass.templateSlot).toMatchSnapshot();
    });

    it('should setParking', () => {
      widgetClass.parkingSlots = [] as HTMLElement[];
      widgetClass.createParkingSlots();
      const firstSlot = document.getElementsByClassName(
        'parking__item',
      )[0] as HTMLDivElement;
      firstSlot.dispatchEvent(new Event('click'));
      expect(firstSlot.classList.contains('selected')).toBe(true);
      expect(widgetClass.setParking).toHaveBeenCalled();
      expect(firstSlot.textContent).toEqual('13.56 EUR');
      // @ts-ignore
      expect(widgetClass).toMatchSnapshot();
    });
  });
});
