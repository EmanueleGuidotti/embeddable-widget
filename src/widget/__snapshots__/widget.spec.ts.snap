// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Widget class Events works properly should setParking 1`] = `
Widget {
  "dictionary": Object {
    "bookYourParking": "Prenota il tuo parcheggio a",
    "featuresTitle": "Sono inclusi i seguenti servizi:",
    "indoorSpace": "Posto auto coperto",
    "insuranceExcluded": "Assicurazione esclusa",
    "insuranceIncluded": "Assicurazione inclusa",
    "outdoorSpace": "Posto auto scoperto",
  },
  "parking": Array [
    Object {
      "id": 123,
      "indoor": false,
      "insurance": false,
      "price": "13.56 EUR",
    },
    Object {
      "id": 456,
      "indoor": false,
      "insurance": true,
      "price": "18.99 EUR",
    },
    Object {
      "id": 243,
      "indoor": true,
      "insurance": true,
      "price": "23.56 EUR",
    },
    Object {
      "id": 700,
      "indoor": true,
      "insurance": false,
      "price": "16.49 EUR",
    },
  ],
  "parkingSlots": Array [
    <div
      class="parking__item selected"
      data-id="123"
      data-indoor="Posto auto scoperto"
      data-insurance="Assicurazione esclusa"
      data-price="13.56 EUR"
    >
      13.56 EUR
    </div>,
    <div
      class="parking__item"
      data-id="456"
      data-indoor="Posto auto scoperto"
      data-insurance="Assicurazione inclusa"
      data-price="18.99 EUR"
    >
      18.99 EUR
    </div>,
    <div
      class="parking__item"
      data-id="243"
      data-indoor="Posto auto coperto"
      data-insurance="Assicurazione inclusa"
      data-price="23.56 EUR"
    >
      23.56 EUR
    </div>,
    <div
      class="parking__item"
      data-id="700"
      data-indoor="Posto auto coperto"
      data-insurance="Assicurazione esclusa"
      data-price="16.49 EUR"
    >
      16.49 EUR
    </div>,
  ],
  "setParking": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "id": 123,
          "indoor": false,
          "insurance": false,
          "price": "13.56 EUR",
        },
      ],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "templateSlot": <div>
    
                    
    <section
      class="vendor"
    >
      
                        
      <div
        class="vendor__toggleable"
      >
        
                            
        <p
          class="vendor__name"
        />
        
                            
        <p
          class="vendor__captureTxt"
        />
        
                            
        <button
          class="vendor__btn"
        >
          open
        </button>
        
                        
      </div>
      
                        
      <div
        class="vendor__features"
      >
        
                            
        <div
          class="features__detail"
        >
          
                                
          <div
            class="vendor__map-container"
          >
            <img
              alt="parking-map"
              class="vendor__map"
            />
          </div>
          
                                
          <div
            class="features__list-container"
          >
            
                                    
            <p
              class="features__title"
            />
            
                                    
            <ul
              class="features__list"
            />
            
                                
          </div>
          
                            
        </div>
        
                        
      </div>
      
                    
    </section>
    
                    
    <section
      class="parking__slots"
    >
      
                    
      <div
        class="parking__item selected"
        data-id="123"
        data-indoor="Posto auto scoperto"
        data-insurance="Assicurazione esclusa"
        data-price="13.56 EUR"
      >
        13.56 EUR
      </div>
      <div
        class="parking__item"
        data-id="456"
        data-indoor="Posto auto scoperto"
        data-insurance="Assicurazione inclusa"
        data-price="18.99 EUR"
      >
        18.99 EUR
      </div>
      <div
        class="parking__item"
        data-id="243"
        data-indoor="Posto auto coperto"
        data-insurance="Assicurazione inclusa"
        data-price="23.56 EUR"
      >
        23.56 EUR
      </div>
      <div
        class="parking__item"
        data-id="700"
        data-indoor="Posto auto coperto"
        data-insurance="Assicurazione esclusa"
        data-price="16.49 EUR"
      >
        16.49 EUR
      </div>
    </section>
  </div>,
  "vendor": Object {
    "departureAirport": "Bergamo Orio al Serio",
    "features": Array [
      "Parcheggio ParkinGo della tipologia selezionata",
      "Navetta da/per aeroporto 24h",
      "Vigilanza e custodia del mezzo 24h",
      "Ripristino gratuito del calo di batteria della tua auto",
    ],
    "map": "http://lorempixel.com/370/390",
    "name": "ParkinGo",
  },
}
`;

exports[`Widget class Events works properly should toggleOpenClose 1`] = `
<div>
  
                    
  <section
    class="vendor"
  >
    
                        
    <div
      class="vendor__toggleable"
    >
      
                            
      <p
        class="vendor__name"
      />
      
                            
      <p
        class="vendor__captureTxt"
      />
      
                            
      <button
        class="vendor__btn"
      >
        close
      </button>
      
                        
    </div>
    
                        
    <div
      class="vendor__features opened"
    >
      
                            
      <div
        class="features__detail"
      >
        
                                
        <div
          class="vendor__map-container"
        >
          <img
            alt="parking-map"
            class="vendor__map"
          />
        </div>
        
                                
        <div
          class="features__list-container"
        >
          
                                    
          <p
            class="features__title"
          />
          
                                    
          <ul
            class="features__list"
          />
          
                                
        </div>
        
                            
      </div>
      
                        
    </div>
    
                    
  </section>
  
                    
  <section
    class="parking__slots"
  >
    
                    
  </section>
</div>
`;

exports[`Widget class should templateSlot created 1`] = `
<div>
  
                    
  <section
    class="vendor"
  >
    
                        
    <div
      class="vendor__toggleable"
    >
      
                            
      <p
        class="vendor__name"
      />
      
                            
      <p
        class="vendor__captureTxt"
      />
      
                            
      <button
        class="vendor__btn"
      >
        open
      </button>
      
                        
    </div>
    
                        
    <div
      class="vendor__features"
    >
      
                            
      <div
        class="features__detail"
      >
        
                                
        <div
          class="vendor__map-container"
        >
          <img
            alt="parking-map"
            class="vendor__map"
          />
        </div>
        
                                
        <div
          class="features__list-container"
        >
          
                                    
          <p
            class="features__title"
          />
          
                                    
          <ul
            class="features__list"
          />
          
                                
        </div>
        
                            
      </div>
      
                        
    </div>
    
                    
  </section>
  
                    
  <section
    class="parking__slots"
  >
    
                    
  </section>
</div>
`;
